# Email Spam or Ham detection 

Aim : Detect the Email whether it's spam or not ?? 


In [1]:
import pandas as pd

In [2]:
main_data = pd.read_csv("E:\DATA-SCIENCE\Project\Email_spam\mail_data.csv")

In [3]:
main_data.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [4]:
main_data['Category'].value_counts()

Category
ham     4825
spam     747
Name: count, dtype: int64

**Binary Encoding**

In [9]:
#Convert the spam as 0 and ham as 1 
main_data.loc[main_data['Category'] == "spam",'Category',] = 0 
main_data.loc[main_data['Category'] == "ham","Category",] = 1 

In [10]:
main_data

Category                                            Message
0           1  Go until jurong point, crazy.. Available only ...
1           1                      Ok lar... Joking wif u oni...
2           0  Free entry in 2 a wkly comp to win FA Cup fina...
3           1  U dun say so early hor... U c already then say...
4           1  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567        0  This is the 2nd time we have tried 2 contact u...
5568        1               Will ü b going to esplanade fr home?
5569        1  Pity, * was in mood for that. So...any other s...
5570        1  The guy did some bitching but I acted like i'd...
5571        1                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [24]:
main_data.loc[:0] #return the 0 index value 

Category                                            Message
0        1  Go until jurong point, crazy.. Available only ...

**Data Cleaning**

In [26]:
main_data.isna().sum() #No Nun data found

Category    0
Message     0
dtype: int64

In [28]:
#Selection of Dependent  and independent variable 
X = main_data['Message']    #independent object
y = main_data['Category']   #dependent or target object

In [31]:
#Split the data into training and testing part
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,y,test_size=0.2,random_state=3) 

In [34]:
X_train.shape,X_test.shape  #4457 for train data vs 1115 for test data

((4457,), (1115,), (4457,))

# Model Development for text data

**here we use TfidfVectorizer library which is combination of Term Frequency and Inverse Document Frequency.**

tf shows us how many time the term(word) are repeated in string, and idf that will count the weight(importance) of that word from document.

In [42]:
from sklearn.feature_extraction.text import TfidfVectorizer as tf

In [43]:
vector_model = tf(min_df=1,stop_words="english",analyzer="word",lowercase=True)    

In [44]:
#let fit-trasform the train data 
X_train_vector = vector_model.fit_transform(X_train)

In [45]:
#let tranform the test data 
X_test_vector = vector_model.transform(X_test)

In [49]:
print(X_train_vector)

  (0, 5413)	0.6198254967574347
  (0, 4456)	0.4168658090846482
  (0, 2224)	0.413103377943378
  (0, 3811)	0.34780165336891333
  (0, 2329)	0.38783870336935383
  (1, 4080)	0.18880584110891163
  (1, 3185)	0.29694482957694585
  (1, 3325)	0.31610586766078863
  (1, 2957)	0.3398297002864083
  (1, 2746)	0.3398297002864083
  (1, 918)	0.22871581159877646
  (1, 1839)	0.2784903590561455
  (1, 2758)	0.3226407885943799
  (1, 2956)	0.33036995955537024
  (1, 1991)	0.33036995955537024
  (1, 3046)	0.2503712792613518
  (1, 3811)	0.17419952275504033
  (2, 407)	0.509272536051008
  (2, 3156)	0.4107239318312698
  (2, 2404)	0.45287711070606745
  (2, 6601)	0.6056811524587518
  (3, 2870)	0.5864269879324768
  (3, 7414)	0.8100020912469564
  (4, 50)	0.23633754072626942
  (4, 5497)	0.15743785051118356
  :	:
  (4454, 4602)	0.2669765732445391
  (4454, 3142)	0.32014451677763156
  (4455, 2247)	0.37052851863170466
  (4455, 2469)	0.35441545511837946
  (4455, 5646)	0.33545678464631296
  (4455, 6810)	0.29731757715898277
  (4

In the context of text analysis and TF-IDF, this could represent the TF-IDF score for a particular term (word) in a document. The row index (0) could correspond to the document, the column index (5413) could correspond to the specific term in the vocabulary, and the value (0.6198254967574347) represents the TF-IDF score for that term in the document.

In [72]:
#for example if we want to check whats inside the term of 5413 at 0 index 
feature_name = vector_model.get_feature_names_out()
term_5413 = feature_name[5413]
print(f"Here the term is:{term_5413}")
print("==================================")

#if we want to check whole term than 
for index,term in enumerate(feature_name):
    print(f"{index} : {term}")

Here the term is:recently
0 : 00
1 : 000
2 : 000pes
3 : 008704050406
4 : 0089
5 : 0121
6 : 01223585236
7 : 01223585334
8 : 0125698789
9 : 02
10 : 0207
11 : 02072069400
12 : 02073162414
13 : 021
14 : 03
15 : 04
16 : 0430
17 : 05
18 : 050703
19 : 0578
20 : 06
21 : 07
22 : 07008009200
23 : 07046744435
24 : 07090201529
25 : 07090298926
26 : 07099833605
27 : 07123456789
28 : 0721072
29 : 07732584351
30 : 07734396839
31 : 07742676969
32 : 07753741225
33 : 0776xxxxxxx
34 : 07781482378
35 : 07786200117
36 : 078
37 : 07808
38 : 07808726822
39 : 07821230901
40 : 07880867867
41 : 0789xxxxxxx
42 : 07946746291
43 : 0796xxxxxx
44 : 07xxxxxxxxx
45 : 08
46 : 0800
47 : 08000407165
48 : 08000776320
49 : 08000839402
50 : 08000930705
51 : 08000938767
52 : 08001950382
53 : 08002888812
54 : 08002986906
55 : 08006344447
56 : 0808
57 : 08081263000
58 : 08081560665
59 : 0825
60 : 0844
61 : 08448350055
62 : 08448714184
63 : 0845
64 : 08450542832
65 : 08452810071
66 : 08452810073
67 : 08452810075over18
68 : 0870

In [52]:
print(X_test_vector)

  (0, 7271)	0.1940327008179069
  (0, 6920)	0.20571591693537986
  (0, 5373)	0.2365698724638063
  (0, 5213)	0.1988547357502182
  (0, 4386)	0.18353336340308998
  (0, 1549)	0.2646498848307188
  (0, 1405)	0.3176863938914351
  (0, 1361)	0.25132445289897426
  (0, 1082)	0.2451068436245027
  (0, 1041)	0.28016206931555726
  (0, 405)	0.2381316303003606
  (0, 306)	0.23975986557206702
  (0, 20)	0.30668032384591537
  (0, 14)	0.26797874471323896
  (0, 9)	0.2852706805264544
  (0, 1)	0.2381316303003606
  (1, 7368)	0.29957800964520975
  (1, 6732)	0.42473488678029325
  (1, 6588)	0.3298937975962767
  (1, 6507)	0.26731535902873493
  (1, 6214)	0.3621564482127515
  (1, 4729)	0.22965776503163893
  (1, 4418)	0.3457696891316818
  (1, 3491)	0.496093956101028
  (2, 7205)	0.22341717215670331
  :	:
  (1110, 3167)	0.5718357066163949
  (1111, 7353)	0.4991205841293424
  (1111, 6787)	0.40050175714278885
  (1111, 6033)	0.4714849709283488
  (1111, 3227)	0.44384935772735523
  (1111, 2440)	0.4137350055985486
  (1112, 7071)

In [67]:
main_data.loc[main_data['Category'] == 1,'Category'].dtype

dtype('O')

In [68]:
#convert the datatype from pandas to numerical
Y_train = Y_train.astype('int')
Y_test = Y_test.astype('int')

In [69]:
Y_train.dtype

dtype('int32')

In [73]:
print(X_train)

3075                  Don know. I did't msg him recently.
1787    Do you know why god created gap between your f...
1614                         Thnx dude. u guys out 2nite?
4304                                      Yup i'm free...
3266    44 7732584351, Do you want a New Nokia 3510i c...
                              ...                        
789     5 Free Top Polyphonic Tones call 087018728737,...
968     What do u want when i come back?.a beautiful n...
1667    Guess who spent all last night phasing in and ...
3321    Eh sorry leh... I din c ur msg. Not sad alread...
1688    Free Top ringtone -sub to weekly ringtone-get ...
Name: Message, Length: 4457, dtype: object


# Model development for Classification 

In [74]:
#We use logistic regression for this data
from sklearn.linear_model import LogisticRegression

In [76]:
Model = LogisticRegression().fit(X_train_vector,Y_train)

**Prediction**

In [84]:
#Prediction on traning data
Pred_train_data = Model.predict(X_train_vector)
print(Pred_train_data)
#Accuracy of traing data 
print(f"Accuracy on traning data:{Model.score(X_train_vector,Y_train)*100:.2f}")

[1 1 1 ... 1 1 0]
Accuracy on traning data:96.70


In [90]:
#Prediction of test data and accuracy score
from sklearn.metrics import accuracy_score
pred_test_data = Model.predict(X_test_vector)
Accuracy_score = accuracy_score(Y_test,pred_test_data)
print(f"Prediction on test data Accuracy score:{Accuracy_score*100:.2f}")

Prediction on test data Accuracy score:96.59


**USER INPUT PREDICTION**

In [92]:
user_input = ['free free free click on link for free money']
input_data = vector_model.transform(user_input)        #transform the user data(test data)
pred_user_data = Model.predict(input_data)

if (pred_user_data[0] == 0):
    print("This Email is SPAM")
else:
    print("This Email is HAM")

This Email is SPAM
